bloomfilter bulk query optimisation with memory prefetching

dcoutts opened the pull request:
For the bloom filter macrobenchmark (on my laptop with 6Mb L3 cache) this is about 2x faster at 25M entries, and about 3x faster at 100M entries.

dcoutts commented on cabal.project:
```diff
@@ -29,6 +29,8 @@ benchmarks: True
 -- Luckily, bloomfilter is not commonly used package, so this is good enough.
 constraints: bloomfilter <0
 
+allow-newer: data-elevator:base
```
No longer necessary now that a new data-elevator is out https://github.com/sgraf812/data-elevator/pull/3 (including the `COMPLETE` pragma to avoid the pattern match warning).

dcoutts commented:
Ok, I think this is properly ready now.

Summary results
* `benchPrepLookups` goes from 6.07 to 2.73, so 2.2x better
* `benchLookupsIO` goes from 18.42 to 16.28, so 1.13x better

So while this reduces the CPU cost quite a bit, the I/O of course still dominates (but use of CPU and I/O resources can overlap so it's still useful).

dcoutts commented on test/Test/Database/LSMTree/Internal/Lookup.hs:
```diff
@@ -113,14 +113,14 @@ prop_bloomQueriesModel ::
     blooms = fmap snd3 runs
     lookupss = concatMap lookups $ getSmallList dats
-    realDefault = bloomQueriesDefault (V.fromList blooms) (V.fromList lookupss)
-    realNonDefault = bloomQueries (V.fromList blooms) (V.fromList lookupss) 10
```
We should find some other way to test it. The other version doesn't have that parameter at all.

jorisdral replied:
If we change `mkTestRun` to not use `AllocFixed 10` by default, then we could either artificially increase the FPR to hit the grow case, or use `Arbitrary` FPR as an argument to the property

dcoutts replied:
So I've addressed this one by checking we get at least coverage of 5% of test cases at a 0.999 FPR.

dcoutts commented on src/Database/LSMTree/Internal/BloomFilterQuery2.hs:
```diff
@@ -0,0 +1,441 @@
+    candidateProbes i_w
+    nextCandidate@(RunIxKeyIx rix kix)
+
+  | i_w < 0x1f && rix < P.sizeofStrictArray filters = do
```
Yes it does. The condition is that once i_r == i_w then we declare the buffer empty and stop. So we can only use 31 of the 32 buffer entries. Hence we must prepare up to 31 at the start.

I agree this could be clearer. I was thinking of moving the buffer empty test to the start of `testCandidateProbe` rather than in `prepNextCandidateProbe`. Or to put it another way, it should be a `while (..) do {..}` loop, not a `do {..} while (..);` loop. This would make the invariants clearer, and should be much the same for code gen/performance.

dcoutts replied:
Added another comment to clarify.

dcoutts commented on src/Database/LSMTree/Internal/BloomFilterQuery2.hs:
```diff
@@ -0,0 +1,441 @@
+    -- assume buff size of 0x20, so mask of 0x1f
+    testCandidateProbe, prepNextCandidateProbe
+      :: Int -> Int
+         -- ^ The read and write indexes within the rolling buffer. This wraps
+         -- around. It is represented with the read index in the least significant
+         -- 8 bits, and the write index in the next higher 8 bits.
```
Yes it is. I was going to do that if I needed more registers, but it turned out that I didn't.

dcoutts replied:
fixed

dcoutts commented on src/Database/LSMTree/Internal/BloomFilterQuery2.hs:
```diff
@@ -0,0 +1,441 @@
+  -> Int
+     -- ^ The write index within the rolling buffer. This wraps around.
+  -> RunIxKeyIx
+     -- ^ The last candidate added to the rolling buffer. When this hits zero
+     -- then no more candidates are added, and we just drain the buffer.
```
Indeed, I flipped the direction in which I go through the candidates but forgot to update the comment.

dcoutts commented on bench/macro/lsm-tree-bench-bloomfilter.hs:
```diff
@@ -235,35 +243,48 @@ elemManyEnv filterSizes rng0 =
+        !khs = V.convert (V.map Bloom.makeHashes ks)
+     in khs `seq` ()
 
 -- | This gives us a combined cost of calculating the series of keys, their
 -- hashes, and then using 'Bloom.elemCheapHashes' with each filter.
```
hmm?

mheinzel replied:
Probably referring to this:

> Technically, this does not include the cost of calculating the keys unless used with benchInBatches

jorisdral replied:
Indeed, I meant "same as https://github.com/IntersectMBO/lsm-tree/pull/325#discussion_r1711218459"

dcoutts commented on bloomfilter/src/Data/BloomFilter/BitVec64.hs:
```diff
@@ -83,6 +98,13 @@ thaw (BV64 bv) = MBV64 <$> P.thaw bv
 roundUpTo64 :: Word64 -> Word64
 roundUpTo64 i = unsafeShiftR (i + 63) 6
 
+remWord64 :: Word64 -> Word64 -> Word64
+#if MIN_VERSION_base(4,17,0)
+remWord64 (W64# x#) (W64# y#) = W64# (x# `remWord64#` y#)
+#else
+remWord64 (W64# x#) (W64# y#) = W64# (x# `remWord#` y#)
+#endif
+
```
They omit the check for 0.
```
instance Integral Word64 where
    ...
    rem     (W64# x#) y@(W64# y#)
        | y /= 0                    = W64# (x# `remWord64#` y#)
        | otherwise                 = divZeroError
```

mheinzel replied:
And we have this in the smart constructor:

```haskell
new hash numBits = MBloom hash numBits' `liftM` V.new numBits'
  where numBits' | numBits == 0                = 1
```

So we should be safe here. Maybe still a good idea to have a check, outside of the hot loop of course.

jorisdral replied:
Makes sense, thanks! I think it's a reasonable to assume bloom filters are non-empty

dcoutts commented on bloomfilter/src/Data/BloomFilter/BitVec64.hs:
```diff
@@ -49,6 +54,13 @@ unsafeIndex (BV64 bv) i =
 unsafeTestBit :: Word64 -> Int -> Bool
 unsafeTestBit w k = w .&. (1 `unsafeShiftL` k) /= 0
 
+{-# INLINE prefetchIndex #-}
+prefetchIndex :: BitVec64 -> Int -> ST s ()
+prefetchIndex (BV64 (P.Vector (I# off#) _ (ByteArray ba#))) (I# i#) =
+    --TODO: explain the shiftR 3 here, vs the shiftR 6 above
+    ST (\s -> case prefetchByteArray0# ba# (off# +# uncheckedIShiftRL# i# 3#) s of
+                s' -> (# s', () #))
```
Yes the whole BitVec64 code only works properly on 64bit systems (and the same assumption is elsewhere in the bloomfilter code).

dcoutts replied:
So i# is an offset in bits, so shiftR 3 to get bytes, and shiftR 3 again to get word64s. And `unsafeIndex` uses `P.unsafeIndex` which works with word64 offsets (and so internally shifts left by 3 again to get a byte offset). So that one effectively does a shiftR 6 and then a shiftL 3. Note that these do not cancel out to give shiftR 3 because that would not zero out the low three bits to give an 8byte aligned address (for the Word64).

For the prefetch we just need a shiftL 3 because we need a byte offset, and we _do not need the byte offset to be 8byte aligned_ (like the `unsafeIndex` does).

jorisdral replied:
Makes sense. And `prefetchByteArray0#` would prefetch a machine word?

dcoutts replied:
Added a comment.

dcoutts commented on src/Database/LSMTree/Internal/StrictArray.hs:
```diff
@@ -0,0 +1,56 @@
+vectorToStrictArray :: forall a. V.Vector a -> StrictArray a
+vectorToStrictArray v =
+    runST $ ST $ \s0 ->
+    case GHC.newArray# (case V.length v of GHC.I# l# -> l#)
+                       (Strict (V.unsafeIndex v 0)) s0 of
```
Yes it could. This is a wrinkle.

dcoutts replied:
I've replaced it with `unsafeCoerce ()` :smile:

dcoutts commented on src/Database/LSMTree/Internal/BloomFilterQuery2.hs:
```diff
@@ -0,0 +1,441 @@
+-- representation for the result: pairs of indexes identifying input Bloom
+-- filters and input keys with a positive test result.
+--
+-- We aim for the number of memory operations in-flight simultaneously to be on
+-- the order of 32. This trades-off memory parallelism with cache use. In
+-- particular this means the algorithm must be able to use a fixed prefetch
+-- \"distance\" rather than this being dependent on the input sizes. To achieve
+-- this, we use a fixed size circular buffer (queue). The buffer size can be
+-- tuned to optimise the prefetch behaviour: indeed we pick exactly 32.
```
I think it's pretty good in general. As I understand it, even decent server class CPUs only support in the range of 10 -- 20 concurrent memory fetches per core.

dcoutts commented on src/Database/LSMTree/Internal/BloomFilterQuery2.hs:
```diff
@@ -0,0 +1,441 @@
+type KeyIx = Int
+type RunIx = Int
+
+-- An implementation of batched bloom filter query, optimised for memory
```
Yes I had thought the same as I was tidying up in haste, and didn't get to it :-)

dcoutts commented on src/Database/LSMTree/Internal/BloomFilterQuery2.hs:
```diff
@@ -0,0 +1,441 @@
+-- We store these in 'PrimArray's as a pair of 64bit words:
+--
+-- * Low 64bit word:
+--   -    padding and the hash number (MSB)
```
Yes and no. We don't need to mask out the low 16bits because we know the upper is empty. I'll see if I can clarify that.

jorisdral replied:
Sure, I just thought it would make it clear that we only use 16 bits for the hash number, since we assert that in `packCandidateProbe`. One might interpret the current comment as hallowing for any number of hash bits + some padding to get to 32 bits in total

dcoutts replied:
clarified

dcoutts commented on src/Database/LSMTree/Internal/BloomFilterQuery2.hs:
```diff
@@ -0,0 +1,441 @@
+type HashNo = Int
+type FilterBitIx = Int
+
+instance P.Prim CandidateProbe where
+    sizeOfType# _ = 16#
+    alignmentOfType# _ = 8#
+
+    indexByteArray# ba i =
+      MkCandidateProbe
+        (P.indexByteArray# ba (indexLo i))
+        (P.indexByteArray# ba (indexHi i))
+    readByteArray# ba i s1 =
+        case P.readByteArray# ba (indexLo i) s1 of { (# s2, lo #) ->
+        case P.readByteArray# ba (indexHi i) s2 of { (# s3, hi #) ->
+        (# s3, MkCandidateProbe lo hi #)
+        }}
+    writeByteArray# ba i (MkCandidateProbe lo hi) s =
+        P.writeByteArray# ba (indexHi i) hi
+       (P.writeByteArray# ba (indexLo i) lo s)
+
+    indexOffAddr# ba i =
+      MkCandidateProbe
+        (P.indexOffAddr# ba (indexLo i))
+        (P.indexOffAddr# ba (indexHi i))
+    readOffAddr# ba i s1 =
+        case P.readOffAddr# ba (indexLo i) s1 of { (# s2, lo #) ->
+        case P.readOffAddr# ba (indexHi i) s2 of { (# s3, hi #) ->
+        (# s3, MkCandidateProbe lo hi #)
+        }}
+    writeOffAddr# ba i (MkCandidateProbe lo hi) s =
+        P.writeOffAddr# ba (indexHi i) hi
+       (P.writeOffAddr# ba (indexLo i) lo s)
```
Oh nice. I didn't know about that.

jorisdral replied:
We've caught a bug before in a `Prim` instance that we defined `lsm-tree` by checking `primLaws`. Often I also add a property test that runs the laws, of which we have an example in the repo: 

https://github.com/IntersectMBO/lsm-tree/blob/9d3d730b6171b34ad12cd90faf3fc6a6b9b6843c/test/Test/Database/LSMTree/Internal/Entry.hs#L30-L37

It just needs a bit of boilerplate, since `quickcheck-classes` exposes `IO` runners instead of `Property` runners

dcoutts replied:
Nice. Done.

dcoutts commented on src/Database/LSMTree/Internal/BloomFilterQuery2.hs:
```diff
@@ -0,0 +1,441 @@
+    let !keyhashes = prepKeyHashes keys
+        !filters'  = P.vectorToStrictArray filters
+    candidateProbes <- P.newPrimArray 0x20
+--  P.setPrimArray 0 0x20 candidateProbes (MkCandidateProbe 0 0)
```
I might leave a comment here that we don't clear/zero the memory, but that this may be helpful for debug.

dcoutts replied:
Added a comment to explain.

dcoutts commented on test/Test/Database/LSMTree/Internal/BloomFilter.hs:
```diff
@@ -62,3 +70,22 @@ prop_total_deserialisation_whitebox hsn (Small len64) =
+        ]
+       ===
+        VU.toList (Bloom1.bloomQueriesDefault (V.fromList filters')
+                                              (V.fromList keys'))
```
Done.

dcoutts commented:
@jorisdral I've addressed most of the review comments (thanks again, there was a lot here!).

I've kept most of the review changes as separate commits (mostly marked FIXUP) to make it clearer what changed. I'll fixup/squash them into their preceding commits before merging.

jorisdral commented on test/Test/Database/LSMTree/Internal/BloomFilter.hs:
```diff
@@ -92,42 +91,81 @@ prop_total_deserialisation_whitebox hsn (Small len64) =
+    -- To get coverage of Bloom2.bloomQueries array resizing we want some
+    -- cases with high FPRs.
+     in tabulate "FPR" [show (round (fpr * 10) * 10 :: Int) ++ "%"] $
+        coverTable "FPR" [("100%", 5)] $
```
Let's add this to `prop_bloomQueries1` as well

jorisdral commented on test/Test/Database/LSMTree/Internal/BloomFilter.hs:
Conditional compilation is causing some compiler errors. You could add separate module just for tests related to `BloomFilterQuery1` and `BloomFilterQuery2`

dcoutts replied:
Yes fixing that now...

jorisdral reviewed (approved):
LGTM

dcoutts commented on lsm-tree.cabal:
```diff
@@ -100,8 +100,17 @@ common language
 
+flag bloom-query-fast
+  description: Use an optimisied Bloom filter query implementation
+  default:     True
+  manual:      True
```
Yes, but the way it's used is that we only use the faster impl for ghc >= 9.4.

We could just make it unconditional (except for ghc version), but then that's harder to benchmark. But perhaps it was only useful during development and we can drop the flag now. Or we can drop it later.

jorisdral replied:
Oh, no sorry, I misunderstood why `manual` is set to `True`. I thought one would not be able to set the flag to `False`, but this should actually be the semantics:
* By default, it is set to `True`
* If one passes `+bloom-query-fast`, it is set to `True` (no-op)
* If one passes `-bloom-query-fast`, it is set to `False`
Is that right?

dcoutts replied:
Correct.

dcoutts commented:
I _think_ this is all ok now. Fixes squashed.
